### PR TITLE
Documentation and unit tests for TypeDict

### PR DESCRIPTION
**Pull request by adisr04**
- Update TypeDict documentation
- Add all unit tests for TypeDict in test_core

### CODE REVIEW

File: relentless/core.py [hunk 1]
```diff
@@ -213,6 +213,11 @@ class TypeDict:
     default
         Initial value to fill in the dictionary, defaults to `None`.
 
+    Raises
+    ------
+    ValueError
```
Comment (mphoward) [inline relentless/core.py:218]: ValueError -> TypeError

File: tests/test_core.py [hunk 2]
```diff
@@ -38,19 +39,106 @@ class test_TypeDict(unittest.TestCase):
...
+        d = relentless.core.TypeDict(types=('A','B'))
+        self.assertEqual(d.types, types)
+
+        #test construction with list input
+        d = relentless.core.TypeDict(types=['A','B'])
+        self.assertEqual(d.types, types)
+
+        #test construction with numpy array input
+        d = relentless.core.TypeDict(types=np.array(['A','B']))
+        self.assertEqual(d.types, types)
```
Comment (mphoward) [inline tests/test_core.py:55]: Drop numpy array test.

File: tests/test_core.py [hunk 3]
```diff
@@ -38,19 +39,106 @@ class test_TypeDict(unittest.TestCase):
 
     def test_init(self):
         """Test construction with different list types."""
-        raise NotImplementedError()
+
```
Comment (mphoward) [inline tests/test_core.py:42]: Don't leave a whitespace between docstrings and code. (You can delete here and below.)

File: tests/test_core.py [hunk 4]
```diff
@@ -38,19 +39,106 @@ class test_TypeDict(unittest.TestCase):
...
+        types = ('A','B')
+
+        #test construction with tuple input
+        d = relentless.core.TypeDict(types=('A','B'))
+        self.assertEqual(d.types, types)
+
+        #test construction with list input
+        d = relentless.core.TypeDict(types=['A','B'])
+        self.assertEqual(d.types, types)
+
```
Comment (mphoward) [inline tests/test_core.py:52]: Can we add a test where the `default` parameter is set to something other than None?

File: relentless/core.py [hunk 5]
```diff
@@ -295,7 +302,7 @@ def __getitem__(self, key):
         key = self._check_key(key)
         return self._data[key]
 
-    def __setitem__(self, key, value):
+    def setitem__(self, key, value):
```
Comment (mphoward) [inline relentless/core.py:305]: This looks like a typo, I doubt the tests work correctly without fixing this.

File: tests/test_core.py [hunk 6]
```diff
@@ -199,26 +196,110 @@ class test_TypeDict(unittest.TestCase):
 
     def test_init(self):
         """Test construction with different list types."""
-        raise NotImplementedError()
+        types = ('A','B')
+        default = {'A':1.0, 'B':1.0}
+
+        #test construction with tuple input
+        d = relentless.core.TypeDict(types=('A','B'))
+        self.assertEqual(d.types, types)
```
Comment (mphoward) [inline tests/test_core.py:204]: Can you put back the checks here that the dict is filled with `None` if `default` is not specified? Also, let's not access `_data` directly. Use either `todict` or access by iterator. You could, for example, do this in a one-liner with list comprehension:
```py
self.assertEqual([d[t] for t in d.types], (None, None))
```

File: tests/test_core.py [hunk 7]
```diff
@@ -199,26 +196,110 @@ class test_TypeDict(unittest.TestCase):
...
+        #test construction with list input
+        d = relentless.core.TypeDict(types=['A','B'])
+        self.assertEqual(d.types, types)
+
+        #test construction with defined default input
+        d = relentless.core.TypeDict(types=('A','B'), default=1.0)
+        self.assertEqual(d.types, types)
+        self.assertEqual(d._data, default)
+
+        types = ('A',)
```
Comment (mphoward) [inline tests/test_core.py:215]: This variable looks out of place here, can you place it with the test below your comment?

File: tests/test_core.py [hunk 8]
```diff
@@ -199,26 +196,110 @@ class test_TypeDict(unittest.TestCase):
...
+        self.assertEqual(d.types, types)
+
+        #test construction with list input
+        d = relentless.core.TypeDict(types=['A','B'])
+        self.assertEqual(d.types, types)
+
+        #test construction with defined default input
+        d = relentless.core.TypeDict(types=('A','B'), default=1.0)
+        self.assertEqual(d.types, types)
+        self.assertEqual(d._data, default)
```
Comment (mphoward) [inline tests/test_core.py:213]: Similarly to above, don't access `_data` directly.
